sycl/doc/design/OptionalDeviceFeatures.md

```diff
@@ -913,211 +913,253 @@ aspects in the array.
-
-The "name" column in this table lists the possible target names.  Since not all
-targets have a corresponding enumerator in the `architecture` enumeration, the
-second column tells when there is such an enumerator.  The last row in this
-table corresponds to all of the architecture names listed in the
-[sycl\_ext\_oneapi\_device\_architecture][8] extension whose name starts with
-`intel_gpu_`.
-
-[8]: <../extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc>
+After the header line, there is one line for each IR module that the
+`sycl-post-link` tool generates.
 
-TODO: This table needs to be filled out for the CPU variants supported by the
-`opencl-aot` tool (avx512, avx2, avx, sse4.2) and for the FPGA targets.  We
-also need to figure out how CUDA fits in here.
+To implement the final filtering pass, the `sycl-post-link` tool uses the
+configuration file to determine the features (e.g. aspects or sub-group sizes)
+that are supported by a target device.  If any IR module uses an incompatible
+feature, the corresponding row in the table is changed to use an empty IR
+module instead.  To illustrate, let's assume that
+`-o=intel_gpu_pvc,output-pvc.table` is passed and that module "mod2.bc" uses
+features that are incompatible with PVC.  The generated table file
+"output-pvc.table" would look like this:
 
-Example of clang compilation invocation with 2 AOT targets and generic SPIR-V:
 ```
-clang++ -fsycl -fsycl-targets=spir64,intel_gpu_12_0_0,ptx64 ...
+[Code|Properties|Symbols]
+mod1.bc|mod1.prop|mod1.sym
+empty.bc|mod2.prop|mod2.sym
+mod3.bc|mod3.prop|mod3.sym
 ```
 
+Notice how the incompatible IR module "mod2.bc" has been replaced with an empty
+IR module "empty.bc".  The other entries on this row ("mod2.prop" and
+"mod2.sym") are unchanged.  It's important to preserve the original properties,
+so that the runtime can diagnose an error if the application mistakenly tries
+to submit a kernel from this IR module to a device that does not support the
+features it uses.
+
+#### AOT compilation and offload wrapping
+
+Currently, the compiler driver invokes the AOT compilation tool (e.g. `ocloc`
+or `opencl-aot`) once for each IR module in the table file that is generated
+by `sycl-post-link`.  When AOT compiling for multiple devices, the driver
+invokes the tool once, passing several target names (e.g.
+`ocloc -device pvc,dg1`).
+
+With the new design, the driver invokes the AOT compilation tool once for each
+IR module in each of the generated file tables.  Each invocation of the tool
+compiles for just a single target device.  To illustrate, let's assume DPC++
+was invoked with `-fsycl-targets=intel_gpu_pvc,intel_gpu_dg1`.  In this case,
+`sycl-post-link` generates two file tables: one for PVC and one for DG1.  The
+driver invokes `ocloc -device pvc` once for each IR module in the PVC file
+table, and it invokes `ocloc -device dg1` once for each IR module in the DG1
+file table.
```

Comment:
In the current toolchain setup, each target device that is provided (i.e. `-fsycl-targets=intel_gpu_pvc,intel_gpu_dg1`) will create an individual unique toolchain for each device from creating the IR -> llvm-link -> sycl-post-link -> clang-offload-wrapper.

Given this, would there still be a need to have multiple `-o` options for `sycl-post-link`?

Reply:
Does this mean that the driver will spawn a separate device compiler "clang" for each of those targets?  (I.e. clang gets invoked once for "intel_gpu_pvc" and again for "intel_gpu_dg1"?)  That would be a compile-time performance regression compared to the current flow.  Currently, if you compile for multiple Intel GPU targets via `-fsycl-targets=spir64_gen -Xs "-device pvc,dg1"`, then clang gets invoked only once.  I thought it would be important to preserve this behavior, so that people don't see a regression in compile-time performance.

Reply:
Okay - this makes sense.  Some additional work would be needed in the driver to accommodate the combining of the targets given multiple intel_gpu* values.  The expected behavior of the `sycl-post-link` call will need to be modified to allow for multiple outputs with the consumption of those outputs to correlate with the new -target values being passed to the individual wrapper calls.

Reply:
> The expected behavior of the `sycl-post-link` call will need to be modified to allow for multiple outputs with the consumption of those outputs to correlate with the new -target values being passed to the individual wrapper calls.

Yes, I agree.  I think this is captured in the sections "Expanded command line options to the post link tool" and "New filtering pass in the post link tool" above.  Or did you mean something more?